Test failures on aarch64-darwin: `RuntimeWarning: divide by zero encountered in equal`

### Description

Running the test suite on aarch64-darwin gives me 3 errors.
```
=================================== FAILURES ===================================
___________________________ LaxRandomTest.test_copy0 ___________________________

self = <random_test.LaxRandomTest testMethod=test_copy0>
make_key = <function key at 0x127b4f400>

    @parameterized.parameters([{'make_key': ctor} for ctor in KEY_CTORS])
    def test_copy(self, make_key):
      key = make_key(8459302)
>     self.assertArraysEqual(key, key.copy())

tests/random_test.py:1713: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
jax/_src/test_util.py:963: in assertArraysEqual
    np.testing.assert_array_equal(x, y, err_msg=err_msg)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (<built-in function eq>, Array((), dtype=key<fry>) overlaying:
[      0 8459302], Array((), dtype=key<fry>) overlaying:
[      0 8459302])
kwds = {'err_msg': '', 'header': 'Arrays are not equal', 'strict': False, 'verbose': True}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           RuntimeWarning: divide by zero encountered in equal

/nix/store/g5cm6iik6p4k39cj9k7a6sg2p09hl7wf-python3-3.10.12/lib/python3.10/contextlib.py:79: RuntimeWarning
_____________________ LaxRandomWithRBGPRNGTest.test_copy0 ______________________

self = <random_test.LaxRandomWithRBGPRNGTest testMethod=test_copy0>
make_key = <function key at 0x127b4f400>

    @parameterized.parameters([{'make_key': ctor} for ctor in KEY_CTORS])
    def test_copy(self, make_key):
      key = make_key(8459302)
>     self.assertArraysEqual(key, key.copy())

tests/random_test.py:1713: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
jax/_src/test_util.py:963: in assertArraysEqual
    np.testing.assert_array_equal(x, y, err_msg=err_msg)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (<built-in function eq>, Array((), dtype=key<rbg>) overlaying:
[      0 8459302       0 8459302], Array((), dtype=key<rbg>) overlaying:
[      0 8459302       0 8459302])
kwds = {'err_msg': '', 'header': 'Arrays are not equal', 'strict': False, 'verbose': True}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           RuntimeWarning: divide by zero encountered in equal

/nix/store/g5cm6iik6p4k39cj9k7a6sg2p09hl7wf-python3-3.10.12/lib/python3.10/contextlib.py:79: RuntimeWarning
__________________ LaxRandomWithUnsafeRBGPRNGTest.test_copy0 ___________________

self = <random_test.LaxRandomWithUnsafeRBGPRNGTest testMethod=test_copy0>
make_key = <function key at 0x127b4f400>

    @parameterized.parameters([{'make_key': ctor} for ctor in KEY_CTORS])
    def test_copy(self, make_key):
      key = make_key(8459302)
>     self.assertArraysEqual(key, key.copy())

tests/random_test.py:1713: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
jax/_src/test_util.py:963: in assertArraysEqual
    np.testing.assert_array_equal(x, y, err_msg=err_msg)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (<built-in function eq>, Array((), dtype=key<urbg>) overlaying:
[      0 8459302       0 8459302], Array((), dtype=key<urbg>) overlaying:
[      0 8459302       0 8459302])
kwds = {'err_msg': '', 'header': 'Arrays are not equal', 'strict': False, 'verbose': True}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           RuntimeWarning: divide by zero encountered in equal

/nix/store/g5cm6iik6p4k39cj9k7a6sg2p09hl7wf-python3-3.10.12/lib/python3.10/contextlib.py:79: RuntimeWarning
=========================== short test summary info ============================
FAILED tests/random_test.py::LaxRandomTest::test_copy0 - RuntimeWarning: divide by zero encountered in equal
FAILED tests/random_test.py::LaxRandomWithRBGPRNGTest::test_copy0 - RuntimeWarning: divide by zero encountered in equal
FAILED tests/random_test.py::LaxRandomWithUnsafeRBGPRNGTest::test_copy0 - RuntimeWarning: divide by zero encountered in equal
== 3 failed, 16331 passed, 2549 skipped, 30 deselected in 8827.06s (2:27:07) ===
```

### What jax/jaxlib version are you using?

jax 0.4.16 (refs/tags/jax-v0.4.16), the jaxlib wheel from PyPI v0.4.16.

### Which accelerator(s) are you using?

CPU

### Additional system info

macOS aarch64-darwin, M1 Pro

### NVIDIA GPU info

_No response_

Just ran again and I'm seeing another failure. Perhaps they're non-deterministic?
```
___________________________ KeyArrayTest.test_async ____________________________

self = <random_test.KeyArrayTest testMethod=test_async>

    def test_async(self):
      key = self.make_keys(10)
    
>     self.assertArraysEqual(key, key.block_until_ready())

tests/random_test.py:2169: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
jax/_src/test_util.py:963: in assertArraysEqual
    np.testing.assert_array_equal(x, y, err_msg=err_msg)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (<built-in function eq>, Array((10,), dtype=key<fry>) overlaying:
[[ 0 28]
 [ 0 29]
 [ 0 30]
 [ 0 31]
 [ 0 32]
 [ 0 33...dtype=key<fry>) overlaying:
[[ 0 28]
 [ 0 29]
 [ 0 30]
 [ 0 31]
 [ 0 32]
 [ 0 33]
 [ 0 34]
 [ 0 35]
 [ 0 36]
 [ 0 37]])
kwds = {'err_msg': '', 'header': 'Arrays are not equal', 'strict': False, 'verbose': True}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           RuntimeWarning: divide by zero encountered in equal

/nix/store/g5cm6iik6p4k39cj9k7a6sg2p09hl7wf-python3-3.10.12/lib/python3.10/contextlib.py:79: RuntimeWarning
```

"ANOTHER ONE" -- DJ Khaled

```
_________________________ KeyArrayTest.test_device_put _________________________

self = <random_test.KeyArrayTest testMethod=test_device_put>

    def test_device_put(self):
      device = jax.devices()[0]
      keys = self.make_keys(4)
      keys_on_device = jax.device_put(keys, device)
>     self.assertArraysEqual(keys, keys_on_device)

tests/random_test.py:2051: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
jax/_src/test_util.py:963: in assertArraysEqual
    np.testing.assert_array_equal(x, y, err_msg=err_msg)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (<built-in function eq>, Array((4,), dtype=key<fry>) overlaying:
[[ 0 28]
 [ 0 29]
 [ 0 30]
 [ 0 31]], Array((4,), dtype=key<fry>) overlaying:
[[ 0 28]
 [ 0 29]
 [ 0 30]
 [ 0 31]])
kwds = {'err_msg': '', 'header': 'Arrays are not equal', 'strict': False, 'verbose': True}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           RuntimeWarning: divide by zero encountered in equal

/nix/store/g5cm6iik6p4k39cj9k7a6sg2p09hl7wf-python3-3.10.12/lib/python3.10/contextlib.py:79: RuntimeWarning
=========================== short test summary info ============================
FAILED tests/random_test.py::KeyArrayTest::test_device_put - RuntimeWarning: divide by zero encountered in equal
==== 1 failed, 1165 passed, 100 skipped, 5 deselected in 267.50s (0:04:27) =====
```

EDIT: ok turns out there's a bunch of these:
```
       > FAILED tests/random_test.py::KeyArrayTest::test_make_array_from_callback - RuntimeWarning: divide by zero encountered in equal
       > FAILED tests/random_test.py::KeyArrayTest::test_make_array_from_single_device_arrays - RuntimeWarning: divide by zero encountered in equal
       > FAILED tests/random_test.py::JnpWithKeyArrayTest::test_array - RuntimeWarning: divide by zero encountered in equal
```

--- a/tests/random_lax_test.py
+++ b/tests/random_lax_test.py
@@ -12,7 +12,6 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import copy
 from functools import partial
 import math
 from unittest import SkipTest
@@ -45,7 +44,6 @@ complex_dtypes = jtu.dtypes.complex
 int_dtypes = jtu.dtypes.all_integer
 uint_dtypes = jtu.dtypes.all_unsigned
 
-KEY_CTORS = [random.key, random.PRNGKey]
 
 @jtu.with_config(jax_legacy_prng_key='allow')
 class LaxRandomTest(jtu.JaxTestCase):
@@ -1165,14 +1163,6 @@ class LaxRandomTest(jtu.JaxTestCase):
 
     for samples in [uncompiled_samples, compiled_samples]:
       self._CheckKolmogorovSmirnovCDF(samples, scipy.stats.lognorm(s=sigma).cdf)
-
-  @parameterized.parameters([{'make_key': ctor} for ctor in KEY_CTORS])
-  def test_copy(self, make_key):
-    key = make_key(8459302)
-    self.assertArraysEqual(key, key.copy())
-    self.assertArraysEqual(key, copy.copy(key))
-    self.assertArraysEqual(key, copy.deepcopy(key))
-    self.assertArraysEqual(key, jax.jit(lambda k: k.copy())(key))
 
 
 threefry_seed = prng_internal.threefry_seed

--- a/tests/random_test.py
+++ b/tests/random_test.py
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import copy
 import enum
 from functools import partial
 import math
@@ -585,6 +586,10 @@ class KeyArrayTest(jtu.JaxTestCase):
   # might also be a more general test of opaque element types. If
   # so, add a corresponding test to to CustomElementTypesTest as well.
 
+  def assertKeysEqual(self, key1, key2):
+    self.assertEqual(key1.dtype, key2.dtype)
+    self.assertArraysEqual(random.key_data(key1), random.key_data(key2))
+
   def test_construction(self):
     key = random.key(42)
     self.assertIsInstance(key, jax_random.PRNGKeyArray)
@@ -661,6 +666,13 @@ class KeyArrayTest(jtu.JaxTestCase):
     self.assertEqual(key.itemsize, key.dtype.itemsize)
     self.assertEqual(key.size, math.prod(key.shape))
     self.assertEqual(key.ndim, len(key.shape))
+
+  def test_key_copy(self):
+    key = self.make_keys()
+    self.assertKeysEqual(key, key.copy())
+    self.assertKeysEqual(key, copy.copy(key))
+    self.assertKeysEqual(key, copy.deepcopy(key))
+    self.assertKeysEqual(key, jax.jit(lambda k: k.copy())(key))
 
   def test_isinstance(self):
     @jax.jit
@@ -916,19 +928,19 @@ class KeyArrayTest(jtu.JaxTestCase):
     device = jax.devices()[0]
     keys = self.make_keys(4)
     keys_on_device = jax.device_put(keys, device)
-    self.assertArraysEqual(keys, keys_on_device)
+    self.assertKeysEqual(keys, keys_on_device)
 
   def test_device_put_sharded(self):
     devices = jax.devices()
     keys = self.make_keys(len(devices))
     keys_on_device = jax.device_put_sharded(list(keys), devices)
-    self.assertArraysEqual(keys, keys_on_device)
+    self.assertKeysEqual(keys, keys_on_device)
 
   def test_device_put_replicated(self):
     devices = jax.devices()
     key = self.make_keys()
     keys_on_device = jax.device_put_replicated(key, devices)
-    self.assertArraysEqual(jnp.broadcast_to(key, keys_on_device.shape), keys_on_device)
+    self.assertKeysEqual(jnp.broadcast_to(key, keys_on_device.shape), keys_on_device)
 
   def test_make_array_from_callback(self):
     devices = jax.devices()
@@ -940,7 +952,7 @@ class KeyArrayTest(jtu.JaxTestCase):
       return jax.vmap(random.key)(i)
     result = jax.make_array_from_callback(shape, sharding, callback)
     expected = jax.vmap(random.key)(jnp.arange(len(devices)))
-    self.assertArraysEqual(result, expected)
+    self.assertKeysEqual(result, expected)
 
   def test_make_array_from_single_device_arrays(self):
     devices = jax.devices()
@@ -950,7 +962,7 @@ class KeyArrayTest(jtu.JaxTestCase):
     keys = random.split(random.key(0), len(devices))
     arrays = [jax.device_put(keys[i:i + 1], device) for i, device in enumerate(devices)]
     result = jax.make_array_from_single_device_arrays(shape, sharding, arrays)
-    self.assertArraysEqual(result, keys)
+    self.assertKeysEqual(result, keys)
 
   def test_key_array_custom_jvp(self):
     def f_raw(x, key):
@@ -1034,7 +1046,7 @@ class KeyArrayTest(jtu.JaxTestCase):
   def test_async(self):
     key = self.make_keys(10)
 
-    self.assertArraysEqual(key, key.block_until_ready())
+    self.assertKeysEqual(key, key.block_until_ready())
     self.assertIsNone(key.copy_to_host_async())
 
   # -- key construction and un/wrapping with impls
@@ -1071,7 +1083,7 @@ class KeyArrayTest(jtu.JaxTestCase):
     k1 = jax.random.key(42, impl=prng_name)
     impl = jax.random.key_impl(k1)
     k2 = jax.random.key(42, impl=impl)
-    self.assertArraysEqual(k1, k2)
+    self.assertKeysEqual(k1, k2)
     self.assertEqual(k1.dtype, k2.dtype)
 
   @jtu.sample_product(prng_name=[name for name, _ in PRNG_IMPLS])
@@ -1082,7 +1094,7 @@ class KeyArrayTest(jtu.JaxTestCase):
     data = jax.random.key_data(k1)
     impl = jax.random.key_impl(k1)
     k2 = jax.random.wrap_key_data(data, impl=impl)
-    self.assertArraysEqual(k1, k2)
+    self.assertKeysEqual(k1, k2)
     self.assertEqual(k1.dtype, k2.dtype)
 
   def test_key_impl_from_string_error(self):
@@ -1136,6 +1148,10 @@ double_threefry_prng_impl = prng_internal.PRNGImpl(
 
 
 class JnpWithKeyArrayTest(jtu.JaxTestCase):
+  def assertKeysEqual(self, key1, key2):
+    self.assertEqual(key1.dtype, key2.dtype)
+    self.assertArraysEqual(random.key_data(key1), random.key_data(key2))
+
   def check_shape(self, func, *args):
     like = lambda keys: jnp.ones(keys.shape)
     out_key = func(*args)
@@ -1257,15 +1273,15 @@ class JnpWithKeyArrayTest(jtu.JaxTestCase):
 
   def test_array(self):
     key = random.key(123)
-    self.assertArraysEqual(key, jnp.array(key))
-    self.assertArraysEqual(key, jnp.asarray(key))
-    self.assertArraysEqual(key, jax.jit(jnp.array)(key))
-    self.assertArraysEqual(key, jax.jit(jnp.asarray)(key))
+    self.assertKeysEqual(key, jnp.array(key))
+    self.assertKeysEqual(key, jnp.asarray(key))
+    self.assertKeysEqual(key, jax.jit(jnp.array)(key))
+    self.assertKeysEqual(key, jax.jit(jnp.asarray)(key))
 
   def test_array_user_dtype(self):
     key = random.key(123)
-    self.assertArraysEqual(key, jnp.array(key, dtype=key.dtype))
-    self.assertArraysEqual(key, jnp.asarray(key, dtype=key.dtype))
+    self.assertKeysEqual(key, jnp.array(key, dtype=key.dtype))
+    self.assertKeysEqual(key, jnp.asarray(key, dtype=key.dtype))
 
   @parameterized.parameters([
     (0,),

